docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,316 @@
+  * Source: https://github.com/dhall-lang/dhall-lang/blob/master/README.md
+  * Golang Library: https://github.com/philandstuff/dhall-golang
+  * Kuberentes Library: https://github.com/dhall-lang/dhall-kubernetes
+* Helm Template
+  * Doc: https://helm.sh/docs/helm/helm_template/
+  * Code: https://github.com/helm/helm/blob/master/cmd/helm/template.go
+
+
+## Summary
+
+Clusterctl handles the lifecycle of Cluster API management and workload
+clusters. It automates fetching the YAML files defining provider components
+and the cluster templates used to define workload clusters. However, as more
+providers contribute to CAPI and the variety of workload cluster templates
+increase, it becomes harder for operators to manage all the cluster template
+flavor permutations. Operators and providers are now leaning towards yaml
+templating tools to properly manage the variety of cluster templates, and to
+simplify the end user experience by limiting the amount of boilerplate YAML
+that operators need to supply.
+
+
+This proposal outlines the need to create an extensible templating mechanism
+within clusterctl so that it can be used effectively with other templating
+tools and continue to provide a seamless user experience.
+
+## Motivation
+
+
+To support a Cluster API that is cloud agnostic, operators need to deploy
+workloads in a multi-cloud environment. Operators and providers may need to
+create and _manage_ multiple cluster templates across a matrix of options and
+configurations. To manage the explosion of yaml manifests, operators may want
+to explore templating tools to minimize duplication and assert type
+correctness.
+
+
+Users looking to deploy workload clusters may be overwhelmed by the number of
+variables needed to configure a workload cluster. Some of these options may be
+circumvented via defaults provided by the operators as part of the templating
+tool.
+
+
+This proposal aims to extend clusterctl’s current templating mechanism of
+simple variable substitution to support other templating formats natively,
+thus benefiting the operators to deploy workload clusters with minimal
+configuration.
+
+### Goals
+
+
+- Clusterctl will not be changing the way it gets values for yaml variables.
+  That is, it will get values from environment variables and the config file.
+- To provide an interface such that other templating solutions may be
+  supported.
+- To implement ytt as the first templating extension.
+
+### Non-Goals
+- To replace existing mechanisms such as kustomize.
+- To replace Clusterctl default mode (environment variables substitution
+  templating).
+
+### Future Work
+- To automatically detect template files for the appropriate template engine.
+- To use the configuration file to determine which provider should use
+  prefered templating mechanisms. For example,
+ ```yaml
+  yamlProcessors:
+  - provider: aws
+    processor: ytt
+ ```
+  We will be leaning towards using a flag for the initial implementation.
```

Comment:
This point is presenting a potential implementation solution, should we just remove the example, or remove it altogether?